File: tests/randomizer.py
import math
import random
from typing import Optional

import expansion.types as types


class Randomizer:
    def __init__(self, seed: int):
        self.seed = seed

    def _apply_seed(self):
        random.seed(self.seed)
        self.seed = int(random.random() * 0xFFFFFFFF)

    def random_value(self, min, max):
        self._apply_seed()
        return random.uniform(min, max)

    def random_position(
            self,
            *,
            rect: Optional[types.Rect] = None,
            center: Optional[types.Position] = None,
            radius: Optional[float] = None,
            min_speed: float = 0,
            max_speed: float = 0) -> types.Position:
        self._apply_seed()
        # Generate random velocity vector
        speed = random.uniform(min_speed, max_speed)
        velocity = types.Vector(x=random.random(), y=random.random())
        velocity.normalize().set_length(speed)
        # Build random position
        if rect:
            assert center is None and radius is None
            return types.Position(
                x=random.uniform(rect.left, rect.right),
                y=random.uniform(rect.bottom, rect.top),
                velocity=velocity
            )
        elif center:
            assert rect is None
            assert radius is not None
            alfa = random.uniform(0, 2 * math.pi)
            r = math.sqrt(random.uniform(0, radius ** 2))
            return types.Position(
                x=center.x + r * math.cos(alfa),
                y=center.y + r * math.sin(alfa),
                velocity=velocity
            )

File: typescript-sdk/tests/helpers/randomizer.ts
import type { Position } from "../../types/common.js";
import type { ResourceItem, ResourceType } from "../../types/resources.js";
import type { Rect } from "./geometry.js";

export function makeResources(
    amounts: Partial<Record<ResourceType, number>>,
): ResourceItem[] {
    return Object.entries(amounts).flatMap(([resource_type, amount]) =>
        amount === undefined
            ? []
            : [{
                resource_type: resource_type as ResourceType,
                amount,
            }],
    );
}

export class Randomizer {
    constructor(private seed: number) {}

    randomValue(min: number, max: number): number {
        return min + (max - min) * this.next();
    }

    randomPosition(options: {
        rect?: Rect;
        center?: Position;
        radius?: number;
        minSpeed?: number;
        maxSpeed?: number;
    }): Position {
        const minSpeed = options.minSpeed ?? 0;
        const maxSpeed = options.maxSpeed ?? 0;
        const speed = this.randomValue(minSpeed, maxSpeed);
        const heading = this.randomValue(0, Math.PI * 2);
        const velocity: [number, number] = [
            speed * Math.cos(heading),
            speed * Math.sin(heading),
        ];

        if (options.rect !== undefined) {
            const { rect } = options;
            return {
                timestamp: 0n,
                point: [
                    this.randomValue(rect.left, rect.right),
                    this.randomValue(rect.bottom, rect.top),
                ],
                velocity,
            };
        }

        if (options.center === undefined || options.radius === undefined) {
            throw new Error(
                "randomPosition requires either rect or center and radius",
            );
        }

        const alfa = this.randomValue(0, Math.PI * 2);
        const r = Math.sqrt(this.randomValue(0, options.radius ** 2));
        return {
            timestamp: 0n,
            point: [
                options.center.point[0] + r * Math.cos(alfa),
                options.center.point[1] + r * Math.sin(alfa),
            ],
            velocity,
        };
    }

    shuffle<T>(items: T[]): T[] {
        for (let i = items.length - 1; i > 0; i -= 1) {
            const j = Math.floor(this.next() * (i + 1));
            const current = items[i]!;
            items[i] = items[j]!;
            items[j] = current;
        }
        return items;
    }

    // mulberry32: deterministic, independent of Math.random().
    private next(): number {
        this.seed |= 0;
        this.seed = (this.seed + 0x6D2B79F5) | 0;
        let t = Math.imul(this.seed ^ (this.seed >>> 15), 1 | this.seed);
        t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
        return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    }
}
